From d4b91d6dc848b6ccf667e243538a5f92ebe55671 Mon Sep 17 00:00:00 2001
From: Matt Madison <matt@madison.systems>
Date: Sun, 10 Dec 2023 09:22:47 -0800
Subject: [PATCH] Makefile: update for OE builds

* Pass needed make variables through from caller
  for the compiler, linker, etc.

* Remove the extra copying of kernel and OOT sources
  where possible (conftest still needs copying, but
  that's small).

* Add KCFLAGS setting to the nvidia-oot make invocation
  to prevent some compilation failures due gcc warning-to-error
  changes in later versions of the compiler

Upstream-Status: Pending
Signed-off-by: Matt Madison <matt@madison.systems>
---
 Makefile | 88 +++++++++++++++++++++++++++++---------------------------
 1 file changed, 45 insertions(+), 43 deletions(-)

diff --git a/Makefile b/Makefile
index f992c51..8d4dc6d 100644
--- a/Makefile
+++ b/Makefile
@@ -1,19 +1,22 @@
 # SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 # SPDX-License-Identifier: BSD-3-Clause
 
-KERNEL_HEADERS ?= /lib/modules/$(shell uname -r)/build
-KERNEL_OUTPUT ?= ${KERNEL_HEADERS}
-
 MAKEFILE_DIR := $(abspath $(shell dirname $(lastword $(MAKEFILE_LIST))))
-NVIDIA_HEADERS ?= ${MAKEFILE_DIR}/out/nvidia-linux-header
-NVIDIA_CONFTEST ?= ${MAKEFILE_DIR}/out/nvidia-conftest
+NVIDIA_CONFTEST ?= $(MAKEFILE_DIR)/out/nvidia-conftest
+HOSTCC ?= gcc
+MODLIB ?= /lib/modules
+CC ?= $(CROSS_COMPILE)gcc
+CXX ?= $(CROSS_COMPILE)g++
+LD ?= $(CROSS_COMPILE)ld.bfd
+AR ?= $(CROSS_COMPILE)ar
+OBJCOPY ?= $(CROSS_COMPILE)objcopy
+
+V ?= 0
 
 ifneq ($(words $(subst :, ,$(MAKEFILE_DIR))), 1)
 $(error source directory cannot contain spaces or colons)
 endif
 
-NPROC ?= $(shell nproc)
-
 .PHONY : help modules modules_install clean conftest nvidia-headers hwpm nvidia-oot nvgpu
 
 # help is default target!
@@ -32,18 +35,21 @@ modules_install: hwpm nvidia-oot nvgpu nvidia-display-install
 clean: hwpm nvidia-oot nvgpu nvidia-display-clean nvidia-dtbs-clean
 
 
-conftest:
+$(NVIDIA_CONFTEST):
+	mkdir -p $@
+
+conftest: | $(NVIDIA_CONFTEST)
 ifeq ($(MAKECMDGOALS), modules)
 	@echo   "================================================================================"
 	@echo   "make $(MAKECMDGOALS) - conftest ..."
 	@echo   "================================================================================"
 	mkdir -p $(NVIDIA_CONFTEST)/nvidia;
-	cp -av $(MAKEFILE_DIR)/nvidia-oot/scripts/conftest/* $(NVIDIA_CONFTEST)/nvidia/;
+	cp -av $(MAKEFILE_DIR)/nvidia-oot/scripts/conftest/* $(NVIDIA_CONFTEST)/nvidia/
 	$(MAKE) ARCH=arm64 \
 		src=$(NVIDIA_CONFTEST)/nvidia obj=$(NVIDIA_CONFTEST)/nvidia \
-		CC=$(CROSS_COMPILE)gcc LD=$(CROSS_COMPILE)ld \
-		NV_KERNEL_SOURCES=$(KERNEL_HEADERS) \
-		NV_KERNEL_OUTPUT=$(KERNEL_OUTPUT) \
+		CC="$(CC)" LD="$(LD)" \
+		NV_KERNEL_SOURCES=$(KERNEL_SRC) \
+		NV_KERNEL_OUTPUT=$(KBUILD_OUTPUT) \
 		-f $(NVIDIA_CONFTEST)/nvidia/Makefile
 endif
 
@@ -55,9 +61,10 @@ hwpm: conftest
 	@echo   "================================================================================"
 	@echo   "make $(MAKECMDGOALS) - hwpm ..."
 	@echo   "================================================================================"
-	$(MAKE) -j $(NPROC) ARCH=arm64 \
-		-C $(KERNEL_HEADERS) \
+	$(MAKE) ARCH=arm64 \
+		-C $(KERNEL_SRC) \
 		M=$(MAKEFILE_DIR)/hwpm/drivers/tegra/hwpm \
+		MODLIB=$(MODLIB) \
 		CONFIG_TEGRA_OOT_MODULE=m \
 		srctree.hwpm=$(MAKEFILE_DIR)/hwpm \
 		srctree.nvconftest=$(NVIDIA_CONFTEST) \
@@ -71,10 +78,12 @@ nvidia-oot: conftest hwpm
 	@echo   "================================================================================"
 	@echo   "make $(MAKECMDGOALS) - nvidia-oot ..."
 	@echo   "================================================================================"
-	$(MAKE) -j $(NPROC) ARCH=arm64 \
-		-C $(KERNEL_HEADERS) \
+	$(MAKE) ARCH=arm64 \
+		-C $(KERNEL_SRC) \
 		M=$(MAKEFILE_DIR)/nvidia-oot \
+		MODLIB=$(MODLIB) \
 		CONFIG_TEGRA_OOT_MODULE=m \
+		KCFLAGS="-Wno-error=address" \
 		srctree.nvidia-oot=$(MAKEFILE_DIR)/nvidia-oot \
 		srctree.hwpm=$(MAKEFILE_DIR)/hwpm \
 		srctree.nvconftest=$(NVIDIA_CONFTEST) \
@@ -89,8 +98,9 @@ nvgpu: conftest nvidia-oot
 	@echo   "================================================================================"
 	@echo   "make $(MAKECMDGOALS) - nvgpu ..."
 	@echo   "================================================================================"
-	$(MAKE) -j $(NPROC) ARCH=arm64 \
-		-C $(KERNEL_HEADERS) \
+	$(MAKE) ARCH=arm64 \
+		-C $(KERNEL_SRC) \
+		MODLIB=$(MODLIB) \
 		M=$(MAKEFILE_DIR)/nvgpu/drivers/gpu/nvgpu \
 		CONFIG_TEGRA_OOT_MODULE=m \
 		srctree.nvidia=$(MAKEFILE_DIR)/nvidia-oot \
@@ -101,30 +111,23 @@ nvgpu: conftest nvidia-oot
 
 
 define display-cmd
-	$(MAKE) -j $(NPROC) ARCH=arm64 TARGET_ARCH=aarch64 \
+	$(MAKE) ARCH=arm64 TARGET_ARCH=aarch64 \
 		-C $(MAKEFILE_DIR)/nvdisplay \
-		LOCALVERSION=$(version) \
-		NV_VERBOSE=0 \
+		NV_VERBOSE=$(V) \
 		KERNELRELEASE="" \
-		SYSSRC=$(NVIDIA_HEADERS) \
-		SYSOUT=$(NVIDIA_HEADERS) \
+		SYSSRC=$(KERNEL_SRC) \
+		SYSOUT=$(KBUILD_OUTPUT) \
+		KBUILD_EXTRA_SYMBOLS=$(MAKEFILE_DIR)/nvidia-oot/Module.symvers \
 		SYSSRCHOST1X=$(MAKEFILE_DIR)/nvidia-oot/drivers/gpu/host1x/include \
-		CC=$(CROSS_COMPILE)gcc \
-		LD=$(CROSS_COMPILE)ld.bfd \
-		AR=$(CROSS_COMPILE)ar \
-		CXX=$(CROSS_COMPILE)g++ \
-		OBJCOPY=$(CROSS_COMPILE)objcopy
+		CC="$(CC)" \
+		LD="$(LD)" \
+		AR="$(AR)" \
+		CXX="$(CXX)" \
+		OBJCOPY="$(OBJCOPY)"
 endef
 
 
-nvidia-headers: nvidia-oot
-	mkdir -p $(NVIDIA_HEADERS) && cp -LR $(KERNEL_HEADERS)/* $(NVIDIA_HEADERS) && \
-		cp -LR $(MAKEFILE_DIR)/nvidia-oot/include/* $(NVIDIA_HEADERS)/include/ && \
-		cat $(KERNEL_HEADERS)/Module.symvers $(MAKEFILE_DIR)/nvidia-oot/Module.symvers > \
-		$(NVIDIA_HEADERS)/Module.symvers
-
-
-nvidia-display: nvidia-headers
+nvidia-display: nvidia-oot
 	@if [ ! -d "$(MAKEFILE_DIR)/nvdisplay" ] ; then \
 		echo "Directory nvdisplay is not found, exiting.."; \
 		false; \
@@ -142,28 +145,27 @@ nvidia-display-install:
 	@echo   "================================================================================"
 	@echo   "make $(MAKECMDGOALS) - nvidia-display ..."
 	@echo   "================================================================================"
-	$(MAKE) -C $(NVIDIA_HEADERS) \
+	$(MAKE) -C $(KERNEL_SRC) \
+		KERNEL_MODLIB=$(MODLIB) \
 		M=$(MAKEFILE_DIR)/nvdisplay/kernel-open modules_install
 
 nvidia-display-clean:
 	@echo   "================================================================================"
 	@echo   "make $(MAKECMDGOALS) - nvidia-display ..."
 	@echo   "================================================================================"
-	if [ -d $(NVIDIA_HEADERS) ]; then \
-		$(display-cmd) clean && \
-		rm -fr $(NVIDIA_HEADERS); \
-	fi
+	$(display-cmd) clean
 
 nvidia-dtbs:
 	@echo   "================================================================================"
 	@echo   "make nvidia-dtbs ..."
 	@echo   "================================================================================"
 	TEGRA_TOP=$(MAKEFILE_DIR) \
-	srctree=$(KERNEL_HEADERS) \
+	srctree=$(KERNEL_SRC) \
 	oottree=$(MAKEFILE_DIR)/nvidia-oot \
-	HOSTCC=gcc \
+	HOSTCC="$(HOSTCC)" \
 	$(MAKE) -f $(MAKEFILE_DIR)/nvidia-oot/scripts/Makefile.build \
 		obj=$(MAKEFILE_DIR)/nvidia-oot/device-tree/platform/generic-dts \
+		objtree=$(KBUILD_OUTPUT) \
 		dtbs
 	@echo   "================================================================================"
 	@echo   "DTBs compiled successfully."
